src/matching_pursuit.jl

```diff
@@ -130,25 +130,36 @@ function sparse_coding(method::Union{MatchingPursuit,ParallelMatchingPursuit,Ort
     return X
 end
 
-@inbounds function matching_pursuit_(
+"""
+OMP implementation. See Pati 1993
+Orthogonal Matching Pursuit: Recursive Function Approximat ion with Applications to Wavelet Decomposition
+(https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=342465)
+"""
+function matching_pursuit_(
     method::Union{OrthogonalMatchingPursuit},
     data::AbstractVector{T}, dictionary::AbstractMatrix{T}, DtD::AbstractMatrix{T};
-    products_init::Union{Nothing,AbstractVector{T}}=nothing) where {T}
+    products_init::Union{Nothing,AbstractVector{T}}=nothing,
+    debug=false) where {T}
     (; max_nnz, max_iter, rtol) = method
 
     n_atoms = size(dictionary, 2)
     residual = copy(data)
     # xdict = DefaultDict{Int, T}(zero(T))
     norm_data = norm(data)
 
-    products = (isnothing(products_init) ? (dictionary' * residual) : products_init)
+    products_init = (isnothing(products_init) ? (dictionary' * residual) : products_init)
+    products = copy(products_init)
     products_abs = abs.(products)  # prealloc
-    A_inv = zeros(T, 0, 0)
+    # A_inv = zeros(T, 0, 0)
+    A = zeros(T, 0, 0)
```

Comment:
Let's check here once more... I know that there are nicer direct QR updates in some paper....

Reply:
Not doing that now. MP with one refit is good enough.